### DISCUSSION TITLE
Extend C Interface?

### DISCUSSION THREAD

**Discussion by pepijndevos** [Xyce/Xyce#18] [Ideas]
Hello,

I'm trying to write a simulation API abstraction layer that I hope can support ngspice, xyce, neuralspice and maybe others.
I have started with ngspice, which has a quite comprehensive C API, which allows amongst other things

- Load a netlist from a string
- Update model and instance parameters
- Run simulations interactively (not present in the spice file) and receive progress updates
- Get direct pointers to simulation outputs

http://ngspice.sourceforge.net/docs/ngspice-manual.pdf 19.3  Shared ngspice API 

In comparison, the current C interface is rather limited, simply wrapping the command line interface to read in a spice file, and then run the simulations contained therein. It must be said that xyce does offer very nice cosimulation facilities, which is absent from ngspice.

I found that the general C++ interface does expose a way to capture simulation results. I think a great feature would be to expose this in the C interface so that it can be used to collect simulation results in Python, C, Rust, Julia, etc. I would be happy to try to implement this, but it's my understanding this is problematic due to legal reasons.

Beyond that, I think there is room to extend the API to offer similar functionality as ngspice offers, but I don't know if there is any interest in supporting that type of programmatic/incremental/interactive uses.

One reason this might be useful is that at least with ngspice, parsing the Sky130 spice library takes several seconds. So during development of smaller simulations it could be beneficial to parse the netlist once and then only update parameters and simulation commands when possible. (this is of course negligible when you run simulations in batch mode that take hours or days to complete)

**Comment by Karlsefni2012** [comment]
Hello, thank you for your interest in Xyce. 

This is a good suggestion.  In general we have been working to improve the various APIs for Xyce, and we may be able to incorporate some of your suggestions.   For example, adding the capability to set parameter values thru a Xyce API is something I think we could add without too much trouble. (if it isn't there already).

One issue is that unlike spice3 and ngspice, we never implemented an interactive mode for Xyce.   This is something we should implement, but at this stage it will take a fair amount of code surgery to  make it happen.  So, the 3rd bullet in your list of capabilities is something that probably can't happen very quickly.  The other stuff probably can happen more quickly, however.

I should note that we've been working recently to expand the general coupling API that you mention above.   That API has had 2 aspects - calling Xyce as a library and also allowing the user to pass models into Xyce.    The calling Xyce as a library part is most directly pertinent to your comment.   It is true that this API is not exposed as part of the default Xyce build, but it is available to anyone who wants to build Xyce from source.  It is available thru an C++ object that is derived from the base Xyce class.

The most recent development on the general interface has been on the model aspect of that interface, to expand it to support a wider variety of model types (including nonlinear) and also set it up to dynamically load python models.    The main driver of that work is to allow people who have developed data-driven models via python libraries such as TensorFlow or pyTorch to run them in Xyce.  That work hasn't made its way to the development branch yet, but it should happen pretty soon, and once it does it will get posted to github the following week.

By the way, we are also making support for the Sky130 PDK a high priority.
